### PR TITLE
PIM-5108: split attribute group selector

### PR DESCRIPTION
**Pull request by sumbobyboys**
The description of what the PR does

| Q                 | A
| ----------------- | ---
| Added Specs       | N
| Added Behats      | Y
| Travis CI is ok   | Y
| Changelog updated | N

In this PR:

- Rework the group selector to be able to use it outside the pef
- Create a save extension to be able to create a generic save button
- Redo the system settings page to avoid using oro configuration stuff
- improvements on behat stability

### CODE REVIEW

File: src/Oro/Bundle/ConfigBundle/Resources/views/Configuration/system.html.twig [hunk 1]
```diff
@@ -1,21 +1,20 @@
...
-        /* global require */
-        require(['oro/form/state'], function (formState) {
+        require(['pim/fetcher-registry', 'pim/form-builder', 'routing'], function (FetcherRegistry, FormBuilder, Routing) {
             'use strict';
-            new formState();
+            FetcherRegistry.initialize().done(function () {
+                $.when(
+                    FormBuilder.build('oro-system-config-form'),
+                    $.get(Routing.generate('oro_config_configuration_system_get'))
+                ).then(function(form, xhr) {
```
Comment (phaseinducer) [inline src/Oro/Bundle/ConfigBundle/Resources/views/Configuration/system.html.twig]: not a huge fan of "xhr" name

**Comment by phaseinducer** [comment]
Awesome to have a generic save and group selector ! Well done =)

File: src/Oro/Bundle/ConfigBundle/Controller/ConfigurationController.php [hunk 2]
```diff
@@ -13,30 +13,8 @@ class ConfigurationController extends Controller
      * @Template()
      * @AclAncestor("oro_config_system")
      */
-    public function systemAction($activeGroup = null, $activeSubGroup = null)
+    public function systemAction()
```
Comment (sumbobyboys) [inline src/Oro/Bundle/ConfigBundle/Controller/ConfigurationController.php]: wil be dropped with the new nav :)

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/common/save.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                this.getRoot().trigger('pim_enrich:form:entity:pre_save');
+                $.ajax({
+                    method: 'POST',
+                    url: this.getSaveUrl(),
+                    contentType: 'application/json',
+                    data: JSON.stringify(this.getFormData())
+                })
+                .then(this.postSave.bind(this))
+                .fail(this.fail.bind(this))
+                .always(function () {
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/common/save.js]: <3 promises

File: src/Oro/Bundle/ConfigBundle/Controller/Rest/ConfigurationController.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+class ConfigurationController
+{
+    /** @var ConfigManager */
+    protected $configManager;
+
+    /** @var array */
+    protected $options;
+
+    /**
+     * @param ConfigManager $configManager
```
Comment (ClementGautier) [inline src/Oro/Bundle/ConfigBundle/Controller/Rest/ConfigurationController.php]: missing phpdoc

File: src/Oro/Bundle/ConfigBundle/Controller/Rest/ConfigurationController.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        $this->configManager = $configManager;
+        $this->options       = $options;
+    }
+
+    /**
+     * Get the current configuration
+     *
+     * @AclAncestor("oro_config_system")
+     */
```
Comment (ClementGautier) [inline src/Oro/Bundle/ConfigBundle/Controller/Rest/ConfigurationController.php]: missing `@return`

File: src/Oro/Bundle/ConfigBundle/Controller/Rest/ConfigurationController.php [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+            ];
+        }
+
+        return new JsonResponse($data);
+    }
+
+    /**
+     * Set the current configuration
+     *
+     * @AclAncestor("oro_config_system")
```
Comment (ClementGautier) [inline src/Oro/Bundle/ConfigBundle/Controller/Rest/ConfigurationController.php]: same

File: src/Oro/Bundle/ConfigBundle/Resources/public/js/form/system/config.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+                return BaseForm.prototype.configure.apply(this, arguments);
+            },
+
+            /**
+             * {@inheritdoc}
+             */
+            render: function () {
+                if (!this.configured) {
+                    return this;
```
Comment (ClementGautier) [inline src/Oro/Bundle/ConfigBundle/Resources/public/js/form/system/config.js]: return void here as it is not fluent if it is configured

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/common/save.js [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+                switch (response.status) {
+                    case 400:
+                        mediator.trigger(
+                            'pim_enrich:form:entity:bad_request',
+                            {'sentData': this.getFormData(), 'response': response.responseJSON}
+                        );
+                        break;
+                    case 500:
+                        /* global console */
+                        console.log('Errors:', response.responseJSON);
```
Comment (ClementGautier) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/common/save.js]: I would not trigger log on the console directly, unless console.log has been overridden somewhere. Logs should be output only on dev environment I think

**Comment by ClementGautier** [comment]
awsome job. GTM after little comments fixed

File: features/Context/DataGridContext.php [hunk 9]
```diff
@@ -774,9 +774,7 @@ public function iRefrestTheGrid()
      */
     public function iClickBackToGrid()
     {
-        $this->wait();
         $this->getSession()->getPage()->clickLink('Back to grid');
-        $this->wait();
```
Comment (phaseinducer) [inline features/Context/DataGridContext.php]: :heart:

File: src/Oro/Bundle/ConfigBundle/Resources/public/js/form/system/config.js (hunk 7 above)
Comment (sumbobyboys) [inline src/Oro/Bundle/ConfigBundle/Resources/public/js/form/system/config.js]: I added a `return this` at the end of the method :)

File: features/Context/Page/Base/Base.php [hunk 10]
```diff
@@ -32,6 +32,23 @@ class Base extends Page
...
 
+
+    /**
+     * Clicks link with specified locator.
+     *
+     * @param string $locator link id, title, text or image alt
+     *
+     * @throws ElementNotFoundException
+     */
+    public function clickLink($locator)
```
Comment (phaseinducer) [inline features/Context/Page/Base/Base.php]: Not sure it's really usefull to refactor this. Our contexts are complicated enough. : /

File: features/Context/DataGridContext.php [hunk 11]
```diff
@@ -774,9 +774,11 @@ public function iRefrestTheGrid()
      */
     public function iClickBackToGrid()
     {
-        $this->wait();
-        $this->getSession()->getPage()->clickLink('Back to grid');
-        $this->wait();
+        $this->spin(function () {
+            $this->getSession()->getPage()->clickLink('Back to grid');
+
```
Comment (jmleroux) [inline features/Context/DataGridContext.php]: You don't need a spin here as you already have one in the spinned method.

File: features/Context/Page/Base/Base.php [hunk 12]
```diff
@@ -377,10 +394,14 @@ public function getDropdownButtonItem($item, $button)
      */
     public function hasSelectedLocale($locale)
     {
-        $selectedLocale = $this->getElement('Locales dropdown')->find('css', 'li.active a');
-        if (null === $selectedLocale) {
-            throw new \Exception('Could not find locales in switcher.');
-        }
+        $selectedLocale = $this->spin(function () {
+            $selectedLocale = $this->getElement('Locales dropdown')->find('css', 'li.active a');
+            if (null === $selectedLocale) {
```
Comment (jmleroux) [inline features/Context/Page/Base/Base.php]: why this ?
better to have the message as spin parameter.

File: features/Context/Page/Base/Base.php [hunk 13]
```diff
@@ -377,10 +394,14 @@ public function getDropdownButtonItem($item, $button)
      */
     public function hasSelectedLocale($locale)
     {
-        $selectedLocale = $this->getElement('Locales dropdown')->find('css', 'li.active a');
-        if (null === $selectedLocale) {
-            throw new \Exception('Could not find locales in switcher.');
-        }
+        $selectedLocale = $this->spin(function () {
```
Comment (phaseinducer) [inline features/Context/Page/Base/Base.php]: Should be :
```
$selectedLocale = $this->spin(function () {
    return $this->getElement('Locales dropdown')->find('css', 'li.active a');
}, 'Could not find locales in switcher.');
```

File: features/Context/DataGridContext.php (hunk 11 above)
Comment (sumbobyboys) [inline features/Context/DataGridContext.php]: nope in fact the clickLink is not spinned

Comment (phaseinducer) [inline features/Context/DataGridContext.php]: :+1:

File: features/Context/Page/Base/Base.php (hunk 13 above)
Comment (sumbobyboys) [inline features/Context/Page/Base/Base.php]: done

File: features/Context/Page/Base/Base.php (hunk 12 above)
Comment (sumbobyboys) [inline features/Context/Page/Base/Base.php]: done

File: features/Context/Page/Base/Base.php (hunk 10 above)
Comment (sumbobyboys) [inline features/Context/Page/Base/Base.php]: Ok, I remove that :)

File: src/Oro/Bundle/ConfigBundle/Resources/public/js/form/system/save.js [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+        'routing',
+        'pim/form/common/save'
+    ],
+    function(
+        module,
+        Routing,
+        SaveForm
+    ) {
+        return SaveForm.extend({
+
```
Comment (phaseinducer) [inline src/Oro/Bundle/ConfigBundle/Resources/public/js/form/system/save.js]: useless blank line

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/common/group-selector.js [hunk 15]
```diff
@@ -0,0 +1,180 @@
+'use strict';
+/**
+ * Attribute group selector extension
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/common/group-selector.js]: Not anymore =)

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/common/group-selector.js [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+             * @return {string}
+             */
+            getCurrent: function () {
+                return sessionStorage.getItem('current_select_group_' + this.code);
+            },
+
+            /**
+             * Set current attribute group
+             *
+             * @param {srting} current
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/common/group-selector.js]: typo

**Comment by phaseinducer** [comment]
GTM

**Comment by jmleroux** [comment]
GTM

**Pull request merged**